tests/components/fritzbox/test_climate.py

```diff
@@ -1,151 +1,306 @@
+    device = FritzDeviceClimateMock()
+    fritz().get_devices.return_value = [device]
+    device.target_temperature = 126.5
+
+    await setup_fritzbox(hass, MOCK_CONFIG)
+    state = hass.states.get(ENTITY_ID)
+    assert state
+    assert state.attributes[ATTR_TEMPERATURE] == 0
+
+
+async def test_update(hass: HomeAssistantType, fritz: Mock):
+    """Test update with error."""
+    device = FritzDeviceClimateMock()
+    fritz().get_devices.return_value = [device]
+
+    await setup_fritzbox(hass, MOCK_CONFIG)
+    state = hass.states.get(ENTITY_ID)
+
+    assert state
+    assert state.attributes[ATTR_CURRENT_TEMPERATURE] == 18
+    assert state.attributes[ATTR_MAX_TEMP] == 28
+    assert state.attributes[ATTR_MIN_TEMP] == 8
+    assert state.attributes[ATTR_TEMPERATURE] == 19.5
+
+    device.actual_temperature = 19
+    device.target_temperature = 20
+
+    next_update = dt_util.utcnow() + timedelta(seconds=200)
+    async_fire_time_changed(hass, next_update)
+    await hass.async_block_till_done()
+    state = hass.states.get(ENTITY_ID)
+
+    assert device.update.call_count == 1
+    assert state
+    assert state.attributes[ATTR_CURRENT_TEMPERATURE] == 19
+    assert state.attributes[ATTR_TEMPERATURE] == 20
+
+
+async def test_update_error(hass: HomeAssistantType, fritz: Mock):
+    """Test update with error."""
+    device = FritzDeviceClimateMock()
+    device.update.side_effect = HTTPError("Boom")
+    fritz().get_devices.return_value = [device]
+
+    await setup_fritzbox(hass, MOCK_CONFIG)
+    assert device.update.call_count == 0
+    assert fritz().login.call_count == 1
+
+    next_update = dt_util.utcnow() + timedelta(seconds=200)
+    async_fire_time_changed(hass, next_update)
+    await hass.async_block_till_done()
+
+    assert device.update.call_count == 1
+    assert fritz().login.call_count == 2
+
+
+async def test_set_temperature_temperature(hass: HomeAssistantType, fritz: Mock):
+    """Test setting temperature by temperature."""
+    device = FritzDeviceClimateMock()
+    fritz().get_devices.return_value = [device]
+
+    await setup_fritzbox(hass, MOCK_CONFIG)
+
+    assert await hass.services.async_call(
+        DOMAIN,
+        SERVICE_SET_TEMPERATURE,
+        {ATTR_ENTITY_ID: ENTITY_ID, ATTR_TEMPERATURE: 123},
+        True,
+    )
+    assert device.set_target_temperature.call_args_list == [call(123)]
+
+
+async def test_set_temperature_mode_off(hass: HomeAssistantType, fritz: Mock):
+    """Test setting temperature by mode."""
+    device = FritzDeviceClimateMock()
+    fritz().get_devices.return_value = [device]
+
+    await setup_fritzbox(hass, MOCK_CONFIG)
+
+    assert await hass.services.async_call(
+        DOMAIN,
+        SERVICE_SET_TEMPERATURE,
+        {
+            ATTR_ENTITY_ID: ENTITY_ID,
+            ATTR_HVAC_MODE: HVAC_MODE_OFF,
+            ATTR_TEMPERATURE: 123,
+        },
+        True,
+    )
+    assert device.set_target_temperature.call_args_list == [call(0)]
```

Comment:
@MartinHjelmare The actual call stack is `set_temperature` -> `set_hvac_mode` -> `set_temperature` when a hvac mode is given. Best test would be to check just the call of `set_hvac_mode` but how? Or is this workflow at all wrong?

Reply:
Always assert the device call at the end, not the entity calls.